cloudvolume/cloudvolume.py

```diff
@@ -156,7 +156,20 @@ def __init__(self, cloudpath, mip=0, bounded=True, autocrop=False, fill_missing=
     except:
       raise Exception("MIP {} has not been generated.".format(self.mip))
 
+  def __setstate__(self, d):
+    """Called when unpickling which is integral to multiprocessing."""
+    self.__dict__ = d 
+
+    if 'cache' in d:
+      self.init_submodules(d['cache'].enabled)
+    else:
+      self.init_submodules(False)
+    
+    # otherwise the pickle might have references to old connections
+    reset_connection_pools() 
```

Comment:
Is calling this global modifier is the right thing to do whenever we unpickle a single cloudvolume instance.
In the case of using a processpool we would be resetting the connection every time. Is that necessary?

Also, I don't know how `reset_connection_pools` work internally for s3 and gcs but is it multithread safe? i.e. when spawning processes that spawns threads (i think it's possible but not 100% sure)?

Reply:
That's a good point that unpickling != process spawning. It's not necessarily appropriate for all cases of unpickling. The code for resetting the spawning pools does this:

```
  S3_POOL = keydefaultdict(lambda service: keydefaultdict(lambda bucket_name: S3ConnectionPool(service, bucket_name)))
  GC_POOL = keydefaultdict(lambda bucket_name: GCloudBucketPool(bucket_name))
```

So it'll keep working no matter what, though it may lead to the accumulation of connections in certain cases (which can then crash the process if it happens enough). It's intended to be used on process spawn as anything in the pool is then invalid and we can ignore the contents with a clean conscience.

Reply:
The trick is to identify when the unpickling is a result of process spawning. Setting the global array to a new pool will cause the connections in the old pool to hang around until `__del__` is called. I don't always trust that to happen.

Reply:
Right, just wanted to make sure that everything inside `S3/GCSConnectionPool` and stuff it calls is also thread safe. 

Did you check out the link to my hacked CV in my comment? lock + a global dict for the pid itself. it's a bit heavy handed so it works so far. Haven't spent too much time to check for other corner cases or more elegant methods though.